File: src/sys/fuzzing/common/artifact.cc
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/sys/fuzzing/common/artifact.h"

#include <lib/syslog/cpp/macros.h>

namespace fuzzing {

Artifact::Artifact(FuzzResult fuzz_result) : empty_(false), fuzz_result_(fuzz_result) {}

Artifact::Artifact(FuzzResult fuzz_result, Input input) : empty_(false), fuzz_result_(fuzz_result) {
  input_ = std::make_unique<Input>(std::move(input));
}

Artifact::Artifact(Artifact&& other) noexcept { *this = std::move(other); }

Artifact& Artifact::operator=(Artifact&& other) noexcept {
  empty_ = other.empty_;
  other.empty_ = true;
  fuzz_result_ = other.fuzz_result_;
  other.fuzz_result_ = FuzzResult::NO_ERRORS;
  input_ = std::move(other.input_);
  return *this;
}

bool Artifact::operator==(const Artifact& other) const {
  if (empty_ && other.empty_) {
    return true;
  }
  if (empty_ || other.empty_) {
    return false;
  }
  if (fuzz_result_ != other.fuzz_result_) {
    return false;
  }
  if (!input_ && !other.input_) {
    return true;
  }
  if (!input_ || !other.input_) {
    return false;
  }
  return *input_ == *other.input_;
}

bool Artifact::operator!=(const Artifact& other) const { return !(*this == other); }

const Input& Artifact::input() const {
  FX_CHECK(!empty_);
  return *input_;
}

Input Artifact::take_input() {
  FX_CHECK(!empty_);
  fuzz_result_ = FuzzResult::NO_ERRORS;
  auto input = std::move(*input_);
  input_.reset();
  empty_ = true;
  return input;
}

Artifact Artifact::Duplicate() const {
  FX_CHECK(!empty_);
  Artifact duplicate(fuzz_result_);
  if (input_) {
    duplicate.input_ = std::make_unique<Input>(input_->Duplicate());
  }
  return duplicate;
}

}  // namespace fuzzing

File: src/sys/fuzzing/fuzzctl/src/artifact.rs
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    crate::input::Input,
    crate::util::digest_path,
    anyhow::{bail, Context as _, Result},
    fidl_fuchsia_fuzzer::{Artifact as FidlArtifact, Result_ as FuzzResult},
    fuchsia_zircon_status as zx,
    std::fs,
    std::path::{Path, PathBuf},
};

/// Combines the results of a long-running fuzzer workflow.
pub struct Artifact {
    /// If `status` is OK, indicates the outcome of the workflow; otherwise undefined.
    pub result: FuzzResult,

    /// The path to which the fuzzer input, if any, has been saved.
    pub path: Option<PathBuf>,
}

impl Artifact {
    /// Returns an artifact for a workflow that completed without producing results or data.
    pub fn ok() -> Self {
        Self { result: FuzzResult::NoErrors, path: None }
    }

    /// Returns an artifact for a workflow that produced a result without data.
    pub fn from_result(result: FuzzResult) -> Self {
        Self { result, path: None }
    }

    /// Returns `artifact.path` as a String, or an empty string if it is `None`.
    pub fn pathname(&self) -> String {
        self.path.as_ref().map(|p| p.to_string_lossy().to_string()).unwrap_or(String::default())
    }
}

/// Reads fuzzer input data from a `FidlArtifact` and saves it locally.
///
/// Returns:
/// Returns an `Artifact` on success. Returns an error if the artifact indicates an error, if
/// it fails to read the data from the `input`, or if it fails to write the data to the file.
///
/// See also `utils::digest_path`.
///
pub async fn save_artifact<P: AsRef<Path>>(
    fidl_artifact: FidlArtifact,
    out_dir: P,
) -> Result<Option<Artifact>> {
    if let Some(e) = fidl_artifact.error {
        if e == zx::Status::PEER_CLOSED.into_raw() {
            return Ok(None);
        }
        bail!("workflow returned an error: ZX_ERR_{}", e);
    }
    let result = fidl_artifact.result.context("invalid FIDL artifact: missing result")?;
    let mut artifact = Artifact::from_result(result);
    if let Some(fidl_input) = fidl_artifact.input {
        let input =
            Input::try_receive(fidl_input).await.context("failed to receive fuzzer input data")?;
        if artifact.result != FuzzResult::NoErrors {
            let path = digest_path(out_dir, Some(artifact.result), &input.data);
            fs::write(&path, input.data).with_context(|| {
                format!("failed to write fuzzer input to '{}'", path.to_string_lossy())
            })?;
            artifact.path = Some(path);
        }
    };
    Ok(Some(artifact))
}

#[cfg(test)]
mod tests {
    use {
        super::save_artifact,
        crate::input::InputPair,
        crate::util::digest_path,
        anyhow::Result,
        fidl_fuchsia_fuzzer::{Artifact as FidlArtifact, Result_ as FuzzResult},
        fuchsia_fuzzctl_test::{verify_saved, Test},
        futures::join,
    };

    #[fuchsia::test]
    async fn test_save_artifact() -> Result<()> {
        let test = Test::try_new()?;
        let saved_dir = test.create_dir("saved")?;

        let input_pair = InputPair::try_from_data(b"data".to_vec())?;
        let (fidl_input, input) = input_pair.as_tuple();
        let send_fut = input.send();
        let fidl_artifact = FidlArtifact {
            result: Some(FuzzResult::Crash),
            input: Some(fidl_input),
            ..FidlArtifact::EMPTY
        };
        let save_fut = save_artifact(fidl_artifact, &saved_dir);
        let results = join!(send_fut, save_fut);
        assert!(results.0.is_ok());
        assert!(results.1.is_ok());
        let saved = digest_path(&saved_dir, Some(FuzzResult::Crash), b"data");
        verify_saved(&saved, b"data")?;
        Ok(())
    }
}
